README.md
<div align="center">

[![template](https://img.shields.io/badge/Repository-template-darkred?style=for-the-badge)](https://github.com/Nakama3942/template_rep)
[![GitHub license](https://img.shields.io/github/license/Nakama3942/mighty_logger?color=gold&style=for-the-badge)](https://github.com/Nakama3942/mighty_logger/blob/master/LICENSE)
[![CHANGELOG](https://img.shields.io/badge/here-CHANGELOG-yellow?style=for-the-badge)](https://github.com/Nakama3942/mighty_logger/blob/master/CHANGELOG.md)
[![PULL_REQUEST_TEMPLATE](https://img.shields.io/badge/here-PULL_REQUEST_TEMPLATE-orange?style=for-the-badge)](https://github.com/Nakama3942/mighty_logger/blob/master/.github/PULL_REQUEST_TEMPLATE.md)

![PyPI](https://img.shields.io/pypi/v/mighty-logger?color=yellow&logo=pypi&logoColor=white&style=for-the-badge)
![GitHub release (latest by date)](https://img.shields.io/github/v/release/Nakama3942/mighty_logger?label=latest%20release&logo=github&style=for-the-badge)
![GitHub commits since latest release (by date including pre-releases)](https://img.shields.io/github/commits-since/Nakama3942/mighty_logger/v0.5.0?include_prereleases&style=for-the-badge)

![PyPI - Python Version](https://img.shields.io/pypi/pyversions/mighty-logger?style=for-the-badge)
![PyPI - Format](https://img.shields.io/pypi/format/mighty-logger?label=PyPI%20format&style=for-the-badge)
![PyPI - Status](https://img.shields.io/pypi/status/mighty-logger?label=PyPI%20status&style=for-the-badge)

![GitHub last commit](https://img.shields.io/github/last-commit/Nakama3942/mighty_logger?style=for-the-badge)
![GitHub Release Date](https://img.shields.io/github/release-date/Nakama3942/mighty_logger?style=for-the-badge)

![GitHub repo size](https://img.shields.io/github/repo-size/Nakama3942/mighty_logger?color=darkgreen&style=for-the-badge)
![GitHub code size in bytes](https://img.shields.io/github/languages/code-size/Nakama3942/mighty_logger?color=darkgreen&style=for-the-badge)
![Lines of code](https://img.shields.io/tokei/lines/github/Nakama3942/mighty_logger?style=for-the-badge)

</div>

# Mighty Logger

### Content

- [Mighty Logger](#mighty-logger)
	- [Content](#content)
	- [Preamble](#preamble)
	- [Overview](#overview)
    - [LICENSE](#license)
	- [Important releases](#important-releases)
	- [Features](#features)
	- [Installation](#installation)
	- [Usage](#usage)
	- [Troubleshooting](#troubleshooting)
	- [Authors](#authors)

## Preamble

> I often came across the opinion that it is better to use not standard output to the console, but full-fledged logging... However, the standard libraries do not provide exactly what I need... Therefore, I decided to make my own library! Which will implement the functionality I need.

I was inspired by the [colored-logs](https://pypi.org/project/colored-logs/) library.

---

*This library has been renamed to "Mighty logger", but this library used to be called "Qt Colored logger". You can search for old commits in this repository, but you can download the old build from the [link](https://pypi.org/project/qt-colored-logger/).*

- [Content](#content)

## Overview

The library implements the formation of a beautifully formatted colored text, similar to a log, which has all the necessary information:

- Device name and registered profile, system name, etc. (this data is displayed only once at the beginning of the logging)
- Log entry time
- Log entry category
- Log entry type
- Entry message

Any information to the output can be turned off (according to the default, everything is included). It is also possible to change the output settings during the logging process. It is possible to change the colors of the foreground text and the background, icons and animations.

- [Content](#content)

## LICENSE

The full text of the license can be found at the following [link](https://github.com/Nakama3942/mighty_logger/blob/master/LICENSE).

> Copyright © 2023 Kalynovsky Valentin. All rights reserved.
>
> Licensed under the Apache License, Version 2.0 (the "License");
> you may not use this file except in compliance with the License.
> You may obtain a copy of the License at
>
>     http://www.apache.org/licenses/LICENSE-2.0
>
> Unless required by applicable law or agreed to in writing, software
> distributed under the License is distributed on an "AS IS" BASIS,
> WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> See the License for the specific language governing permissions and

- [Content](#content)

## Important releases

<details><summary>See the important releases</summary>

- [x] v0.0.1 - Alpha-release (the very first version of the simplest Logger has been published)
- [x] v0.0.2 - Little update (added multiple entry types and colors)
- [x] v0.0.3 - Types update (added even more multiple entry types and colors)
- [x] v0.0.4 - Color update (added the entire X11 color table and reworked the color system)
- [x] v0.1.0 - "First release" update (complete basic HTML logger)
- [x] v0.2.0 - Structural update (added basic console logger with HTML base)
- [x] v0.2.1 - Protections update (hidden part of the functionality)
- [x] v0.3.0 - Background update (added background for log entries)
- [x] v0.4.0 - Buffer update (added text buffer)
- [x] v0.5.0 - Unifying update (console and HTML are combined into one class)
- [x] v0.5.1 - Hints update (added hint symbols (icons) near log entries status)
- [x] v0.6.0 - Progress update (added start of some log entries in threads (process))
- [x] v0.6.1 - Animation update (extended animations in processes)
- [x] v0.7.0 - "Buffer improvement" update (buffer development completed)
- [x] v0.7.1 - Modding update (buffer modification added - sorting, searching and selecting)
- [x] v0.7.2 - Categories update (separated the entry category from the type)
- [x] v0.8.0 - Export update (added conversion to csv)
- [x] v0.9.0 - Extension update (made wheel format and instruction of toml)
- [x] v0.9.1 - Documenting update (documented library)
- [x] v0.9.2 - Feature update (made optimizations)
- [x] v0.9.3 - Web docs update (added generation of web docs)
- [x] v1.0.0 - Completion of logger development (logger development completed)
- [ ] v1.1.0 - Font update (added a class that formats text outside the logger)

</details>

- [Content](#content)

## Features

- **Support for a large number of logger entry types** (empty, entry: debug, debug_performance, performance, event, audit, metrics, user, message, info, notice, warning, error, critical, resolved, unresolved)
- **Support for five working environments** (CONSOLE, PLAIN_CONSOLE, HTML, MARKDOWN, PLAIN)
- **Provides publisher methods**
- **Provides to sort the logs**
- **Provides to search the logs**
- **Provides to select the desired logs**
- **Provides to export logs to csv**
- **Implemented own Text Buffer**
- **Provided the ability to control the Buffer**
- **Logs can be saved and loaded**
- **Implemented wrapper method for Python input()**
- **You can run logging Processes**
- **You can run logging Timers**
- **Implemented a simplified Logger class**
- **When creating a Logger, an entry is created about the characteristics of the system**
- **Show time of logging for every entry log**
- **Provided are more than 20 animations for indefinite processes and 6 for definite processes**
- **Provided are ANSI escape code table**
- **Provides colors in RGB, HEX, ANSI**
- **Provided are 4 sorting options** (by time, by reverse time, by category, by type)
- **Provided are 5 exceptions**

## Installation

To install the library, enter the command:

```sh
pip install mighty_logger
```

## Usage

This is the simplest example of using the library:

```python
from mighty_logger import Logger

if __name__ == "__main__":
	logger = Logger()
	logger.message("Hello world!")
	print("\n".join(logger.might.buffer))
```

The outputs in console will contain the following text:

<pre>
-Unknown?entry> $DESKTOP-8KG0R64:User:Windows:10.0.19045:64bit:WindowsPE:AMD64
-?entry>          *2023-07-04 21:12:05.919593 📝 #STATUS: @MESSAGE - Hello world!
</pre>

- [Content](#content)

## Troubleshooting

All functionality of the library has been tested by me, but if you have problems using it, the code does not work, have suggestions for optimization or advice for improving the style of the code and the name - I invite you [here](https://github.com/Nakama3942/mighty_logger/blob/master/CONTRIBUTING.md) and [here](https://github.com/Nakama3942/mighty_logger/blob/master/CODE_OF_CONDUCT.md).

- [Content](#content)

## Authors

<table align="center" style="border-width: 10; border-style: ridge">
	<tr>
		<td align="center"><a href="https://github.com/Nakama3942"><img src="https://avatars.githubusercontent.com/u/73797846?s=400&u=a9b7688ac521d739825d7003a5bd599aab74cb76&v=4" width="150px;" alt=""/><br /><sub><b>Kalynovsky Valentin</b></sub></a><sub><br />"Ideological inspirer and Author"</sub></td>
		<!--<td></td>-->
	</tr>
<!--
	<tr>
		<td></td>
		<td></td>
	</tr>
-->
</table>


CHANGELOG.md
# Changelog
<!--
## vX.X.X (DATE)

#### Bug Fixes: (Исправления ошибок и неисправностей)
- [# XXX](https : / / github . com / XXX) DESCRIPTION

#### Documenting: (Документирование функционала)
- [# XXX](https : / / github . com / XXX) DESCRIPTION

#### Enhancements: (Улучшения функционала: добавление, изменение)
- [# XXX](https : / / github . com / XXX) DESCRIPTION

#### Deprecated: (Отмеченные устаревшие/удалённые функции/методы)
- [# XXX](https : / / github . com / XXX) DESCRIPTION

#### Security: (Изменения, связанные с безопасностью или исправлением уязвимостей)
- [# XXX](https : / / github . com / XXX) DESCRIPTION

#### Performance: (Изменения, направленные на улучшение производительности)
- [# XXX](https : / / github . com / XXX) DESCRIPTION

#### Refactoring: (Переименование, перестройка или оптимизация кода без изменения функциональности)
- [# XXX](https : / / github . com / XXX) DESCRIPTION

---
-->

## Alpha-release v0.0.1 (13.03.2023)

#### Release
The library implements the formation of a beautifully formatted colored text, similar to a log, which has all the necessary information:
- Logging time;
- Name of device and profile that logged;
- Log status;
- Description of the log status;
- Log type;
- Log message.

Any information to the output can be turned off (according to the standard, everything is included). It is also possible to change the output settings during the logging process. It is possible to change colors (class `PickerModifierQ`).

*!!!ATTEMPTION!!! At the moment, logging is implemented only in the form of HTML code for QTextBrowser for PyQt, since quite often I need to output the log not to the console, but to the program and save it to a file, including saving colors. Therefore, in this version, output to the console is not implemented, but only in QTextBrowser, however, in the next versions, a lot of functionality will be implemented for easy and convenient logging!*

---

## Little update v0.0.2 (16.03.2023)

#### Bug Fixes:
- Fixed some typos.

#### Enhancements:
- Added new colors:
	- OCEANBLUE;
	- DARKOCEANBLUE;
- Changed color names:
	- CYAN -> BLUE;
	- DARKCYAN -> DARKBLUE;
- Added an ID to each logger class;
- Added new methods to the Logger class:
	- `SUCCESS()`;
	- `FAIL()`;
	- `START_PROCESS()`; *stub - not implemented*
	- `STOP_PROCESS()`; *stub - not implemented*
- Added some links to PyPi.

---

## Types update v0.0.3 (17.03.2023)

#### Documenting:
- The `LoggerQ` class is fully documented.

#### Enhancements:
- Added new types of log output:
	1. `DEBUG_PERFORMANCE()`;
	2. `PERFORMANCE()`;
	3. `EVENT()`;
	4. `AUDIT()`;
	5. `METRICS()`;
	6. `USER()`;
	7. `MESSAGE()`;
	8. `NOTICE()`;
- Added new colors:
	1.  FIREBRICK *replaced RED*
	2.  MEDIUMSPRINGGREEN
	3.  SPRINGGREEN
	4.  MEDIUMSEAGREEN
	5.  SEAGREEN
	6.  FORESTGREEN *not used yet*
	7.  YELLOWGREEN
	8.  OLIVEDRAB
	9.  OLIVE
	10. DARKOLIVEGREEN
	11. AQUAMARINE *replaced BLUE*
	12. TURQUOISE *replaced DARKBLUE*
	13. SKYBLUE *replaced OCEANBLUE*
	14. LIGHTSKYBLUE *replaced DARKOCEANBLUE*
	15. BLUE *Adopted its color according to the X11 standards table*
	16. MEDIUMBLUE
	17. DARKBLUE *Adopted its color according to the X11 standards table*
	18. NAVY
	19. BLUEVIOLET *replaced VIOLET*
	20. DARKVIOLET *Adopted its color according to the X11 standards table*
	21. GAINSBORO
	22. LIGHTGREY
	23. SILVER
	24. DIMGREY

---

## Color update v0.0.4 (25.03.2023)

#### Bug Fixes:
- Found and fixed several minor bugs.

#### Documenting:
- The `HtmlColorSetInitQ` class is fully documented;
- Slightly fixed CHANGELOG.

#### Enhancements:
- The color system has been completely redesigned (i.e. the system for working with colors);
- Added ALL X11 table colors [from this page](https://en.wikipedia.org/wiki/Web_colors);
- Added functions for getting color values in different formats by their names;
- Added `ColorException`;
- Started work with console logger (*not yet available*);
- Now only one object of the `HtmlColorSetInitQ` class can be created;
- Now only one object of the `LoggerQ` class can be created;
- The color table of the LoggerQ logger has been changed - now the table does not store color names, but abstract names:
	- TIME;
	- USER;
	- STATUS;
	- STATUS_MESSAGE;
	- TYPE_DEBUG;
	- DEBUG_MESSAGE;
	- TYPE_DEBUG_PERFORMANCE;
	- DEBUG_PERFORMANCE_MESSAGE;
	- TYPE_PERFORMANCE;
	- PERFORMANCE_MESSAGE;
	- TYPE_EVENT;
	- EVENT_MESSAGE;
	- TYPE_AUDIT;
	- AUDIT_MESSAGE;
	- TYPE_METRICS;
	- METRICS_MESSAGE;
	- TYPE_USER;
	- USER_MESSAGE;
	- TYPE_MESSAGE;
	- MESSAGE_MESSAGE;
	- TYPE_INFO;
	- INFO_MESSAGE;
	- TYPE_NOTICE;
	- NOTICE_MESSAGE;
	- TYPE_WARNING;
	- WARNING_MESSAGE;
	- TYPE_ERROR;
	- ERROR_MESSAGE;
	- TYPE_CRITICAL;
	- CRITICAL_MESSAGE;
	- TYPE_PROGRESS;
	- PROGRESS_MESSAGE;
	- TYPE_SUCCESS;
	- SUCCESS_MESSAGE;
	- TYPE_FAIL;
	- FAIL_MESSAGE;
- Now, initially this table is not initialized and the logger will not be able to work. First you need to initialize the table by creating an object of the `HtmlColorSetInitQ` class, which also provides methods for changing this very logger color table.

#### Deprecated:
- Implemented library functionality tests for early detection of minor bugs.

#### Refactoring:
- `PickerModifierQ` class renamed to `HtmlColorSetInitQ`.

---

## "First release" update v0.1.0 (26.03.2023)

#### Bug Fixes:
- Fixed a bug that occurs when passing arguments to the `LoggerQ` constructor;
- Fixed a bug due to which the color value was not correctly saved in the logger's color table when changing the color manually.

#### Documenting:
- Completed documentation of the library;
- Wrote the actual README (which corresponds to the current version).

#### Enhancements:
- Slightly redefined colors in the logger;
- Changed the order of colors in the color table.

---

## Structural update v0.2.0 (29.03.2023)

#### Documentation:
- Documented all new functionality;
- Fixed typos and outdated information in old documentation.

#### Enhancements:

mighty_logger/basic/sort.py
"""
...
\n
Copyright © 2023 Kalynovsky Valentin. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

import re, os

template = [
	"--------------------------------------------------------------------------------",
	"Sorted logs",
	"--------------------------------------------------------------------------------",
	"Unsorted empty entries"
]

category_order = {
	"%": 0,
	"~": 1,
	"@": 2,
	"!": 3,
	"&": 4,
	"^": 5
}

type_order = {
	'DEBUG': 0,
	'DEBUG-PERFORMANCE': 1,
	'PERFORMANCE': 2,
	'EVENT': 3,
	'AUDIT': 4,
	'METRICS': 5,
	'USER': 6,
	'MESSAGE': 7,
	'INFO': 8,
	'NOTICE': 9,
	'WARNING': 10,
	'!ERROR': 11,
	'!!@CRITICAL': 12,
	'RESOLVED': 13,
	'UNRESOLVED': 14,
	'INITIATION': 15,
	'PROGRESS': 16,
	'ACHIEVEMENT': 17,
	'MILESTONE': 18,
	'SUCCESS': 19,
	'FAIL': 20,
	'START-TIMER': 21,
	'TIMER-MARK': 22,
	'STOP-TIMER': 23
}

class SortingKeyType:
	def __init__(self, sorting_key: int):
		self.__sorting_key: int = sorting_key

	@property
	def sorting_key(self) -> int:
		return self.__sorting_key

class SortingKeys:
	SORT_ON_TIME = SortingKeyType(0)
	SORT_ON_TIME_WITH_REVERSE = SortingKeyType(1)
	SORT_ON_CATEGORY = SortingKeyType(2)
	SORT_ON_TYPE = SortingKeyType(3)

def sort(entries: list[str], key: SortingKeyType = SortingKeys.SORT_ON_TYPE) -> list:
	sorting_entries = []
	count_separators = 0

	for index, entry in enumerate(entries[1:]):
		cleared_entry = re.sub(r"\033\[.*?m", "", entry)
		if cleared_entry.startswith("-?"):
			parts = cleared_entry.split()
			sorting_entries.append([entries.pop(index + 1 - len(sorting_entries) - count_separators), " ".join(parts[1:3]), parts[4][0], parts[4][1:]])
			continue
		if cleared_entry.startswith("--"):
			entries.pop(index + 1 - len(sorting_entries) - count_separators)
			count_separators += 1
			continue

	match key:
		case SortingKeys.SORT_ON_TIME:
			sorting_entries.sort(key=lambda x: x[1])
		case SortingKeys.SORT_ON_TIME_WITH_REVERSE:
			sorting_entries.sort(key=lambda x: x[1], reverse=True)
		case SortingKeys.SORT_ON_CATEGORY:
			sorting_entries.sort(key=lambda x: category_order.get(x[2], float("inf")))
		case SortingKeys.SORT_ON_TYPE:
			sorting_entries.sort(key=lambda x: type_order.get(x[3], float("inf")))

	sorted_entries = [item[0] for item in sorting_entries]

	entries[1:1] = template
	entries[3:3] = sorted_entries

	return entries

if __name__ == "__main__":
	old_logs = [
		"\033[38;2;255;215;0m-Installer?entry> $DESKTOP-8KG0R64:User:Windows:10.0.19045:64bit:WindowsPE:AMD64\033[0m",
		"--------------------------------------------------------------------------------",
		"Enter password: 1234",
		"\033[38;2;139;0;0m-?entry>          \033[38;2;218;112;214m*2023-06-26 21:00:53.276473 🚫 \033[38;2;178;34;34m!!ERROR - \033[38;2;139;0;0mIncompatibility found\033[0m",
		"\033[38;2;139;0;0m-?entry>          \033[38;2;218;112;214m*2023-06-26 21:00:54.277457 🚫 \033[38;2;178;34;34m&FAIL - \033[38;2;139;0;0mProgram not installed\033[0m",
		"1234",
		"\033[38;2;210;180;140m-?entry>          \033[38;2;218;112;214m*2023-06-26 21:00:59.284448 🐞 \033[38;2;222;184;135m%DEBUG - \033[38;2;210;180;140mbla bla bla\033[0m",
		"\033[38;2;210;180;140m-?entry>          \033[38;2;218;112;214m*2023-06-26 21:01:01.287537 🐞 \033[38;2;222;184;135m%DEBUG - \033[38;2;210;180;140mString has deleted\033[0m",
		"\033[38;2;176;224;230m-?entry>          \033[38;2;218;112;214m*2023-06-26 21:00:46.818626 📝 \033[38;2;175;238;238m@MESSAGE - \033[38;2;176;224;230mProgram installation started\033[0m",
	]
	print("\n".join(old_logs))
	new_logs = sort(old_logs, SortingKeys.SORT_ON_TIME)
	print("\n".join(new_logs))
